src/mango/src/mango_cursor_view.erl

```diff
@@ -457,12 +480,21 @@ handle_message({row, Props}, Cursor) ->
             couch_log:error("~s :: Error loading doc: ~p", [?MODULE, Error]),
             {ok, Cursor}
     end;
-handle_message({execution_stats, ShardStats}, #cursor{execution_stats = Stats} = Cursor) ->
-    {docs_examined, DocsExamined} = ShardStats,
-    Cursor1 = Cursor#cursor{
+% TODO: remove clause couchdb 4 -- mixed-version upgrade support
+handle_message(
+    {execution_stats, {docs_examined, DocsExamined}}, #cursor{execution_stats = Stats} = Cursor0
```

Comment:
Thanks Nick for the comments!  I have had the reasons for these choices, let me add them for clarification:

- The `{execution_stats, {docs_examined, DocsExamined}}` pattern was embedded into the function head to avoid the potential overlap with other clause.  My understanding is that heads are checked top-down and the body of the first matching one is going to be evaluated.  If there is no direct pattern to match on the contents of the argument of `execution_stats` (i.e. the second element in the tuple) in the head, the clause will be skipped and missed.

-  For the second clause on `execution_stats`, I avoided to match on the type of `ShardStats` to keep it opaque.  Ideally, `mango_cursor_view:shard_stats_get/2` will handle the map check eventually.  I am aware that might be too late, but there are no other alternatives present to be confused with.  At the same time, changes to the format of shard stats is encapsulated into the implementation of the `shard_stats/2` and `shard_stats_get/2` functions.

Reply:
> The {execution_stats, {docs_examined, DocsExamined}} pattern was embedded into the function head to avoid the potential overlap with other clause. 

That's a good way to do it. The heads should go from the most specific one to the more general.  The first comment wasn't about the order of clauses but about not having the line get too long. In the first head we're doing at least two unrelated things: match on `{execution_stats, {docs_examined, DocsExamined}}` and picking out `Stats` out of the `#cursor`. The idea of the cleanup is to pick out the `Stats` variable in a separate line.

> I avoided to match on the type of ShardStats to keep it opaque. 

I had noticed but it would still look better have an extra match since we're juggling two different formats. We're not breaking the opaqueness too much as we're not picking out individual keys from the map.

Reply:
> The idea of the cleanup is to pick out the `Stats` variable in a separate line.

Oh, you are right!  Yes, that is a good idea.

> We're not breaking the opaqueness too much as we're not picking out individual keys from the map.

That one still forces the map type on the container.  That is true that checking only if that is map would not interfere with the actual keys, but what if we wanted to move away from the maps for a different representation?

Reply:
It's a trade-off: preserve complete opaqueness or provide immediate visual and runtime feedback that the type is wrong in that clause. If we switch away from a map, we'd get an immediate failure in testing.  In a strongly-typed language it might be better not to be "leaky",  but in Erlang matching on things we expect as early as possible is more common. But that's fine, if you feel strongly about. It's +1 either, way.  If you rebase again, I'll merge (to get a clean rebase in main).

Reply:
I do not have a strong feeling about it.  As you have also put it, this might be kind of a habit that I acquired from strongly typed languages and I accept that the same may apply to Erlang.  I can follow this idiom.

But please do not merge this PR for now.  In response to this approach, I was suggested to handle the question of transition through configuration options and add support for both the sender and the receiver sides in a single change.  I will file another PR for that alternative, and it is very likely that will be the winner.